lib/srv/desktop/deskproto/proto.go

```diff
@@ -0,0 +1,254 @@
+
+	buf := new(bytes.Buffer)
+	if err := binary.Write(buf, binary.BigEndian, header{
+		Type:   byte(TypePNGFrame),
+		Left:   uint32(f.Img.Bounds().Min.X),
+		Top:    uint32(f.Img.Bounds().Min.Y),
+		Right:  uint32(f.Img.Bounds().Max.X),
+		Bottom: uint32(f.Img.Bounds().Max.Y),
+	}); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := png.Encode(buf, f.Img); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return buf.Bytes(), nil
+}
+
+func decodePNGFrame(buf []byte) (PNGFrame, error) {
+	var header struct {
+		Left, Top     uint32
+		Right, Bottom uint32
+	}
+	r := bytes.NewReader(buf[1:])
+	if err := binary.Read(r, binary.BigEndian, &header); err != nil {
+		return PNGFrame{}, trace.Wrap(err)
+	}
+	img, err := png.Decode(r)
+	if err != nil {
+		return PNGFrame{}, trace.Wrap(err)
+	}
+	// PNG encoding does not preserve offset image bounds.
+	// Opportunistically restore them based on the header.
+	switch img := img.(type) {
+	case *image.RGBA:
+		img.Rect = image.Rect(int(header.Left), int(header.Top), int(header.Right), int(header.Bottom))
+	case *image.NRGBA:
+		img.Rect = image.Rect(int(header.Left), int(header.Top), int(header.Right), int(header.Bottom))
+	}
+	return PNGFrame{Img: img}, nil
+}
+
+// MouseMove is the mouse movement message.
+// https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md#3---mouse-move
+type MouseMove struct {
+	X, Y uint32
+}
+
+func (m MouseMove) Encode() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(byte(TypeMouseMove))
+	if err := binary.Write(buf, binary.BigEndian, m); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return buf.Bytes(), nil
+}
+
+func decodeMouseMove(buf []byte) (MouseMove, error) {
+	var m MouseMove
+	err := binary.Read(bytes.NewReader(buf[1:]), binary.BigEndian, &m)
+	return m, trace.Wrap(err)
+}
+
+// MouseButtonType identifies a specific button on the mouse.
+type MouseButtonType byte
+
+const (
+	LeftMouseButton   = MouseButtonType(0)
+	MiddleMouseButton = MouseButtonType(1)
+	RightMouseButton  = MouseButtonType(2)
+)
+
+// ButtonState is the press state of a keyboard or mouse button.
+type ButtonState byte
+
+const (
+	ButtonNotPressed = ButtonState(0)
+	ButtonPressed    = ButtonState(1)
+)
+
+// MouseButton is the mouse button press message.
+// https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md#4---mouse-button
+type MouseButton struct {
+	Button MouseButtonType
+	State  ButtonState
+}
+
+func (m MouseButton) Encode() ([]byte, error) {
+	return []byte{byte(TypeMouseButton), byte(m.Button), byte(m.State)}, nil
+}
+
+func decodeMouseButton(buf []byte) (MouseButton, error) {
+	var m MouseButton
+	err := binary.Read(bytes.NewReader(buf[1:]), binary.BigEndian, &m)
+	return m, trace.Wrap(err)
+}
+
+// KeyboardButton is the keyboard button press message.
+// https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md#4---keyboard-input
+type KeyboardButton struct {
```

Comment:
I'm curious - how are multi-key combinations (i.e. `Ctrl-Alt-Del`) are represented in the protocol?

Reply:
As one message per key press - separate `ctrl`/`alt`/`del`.
Each message represents key going up or down only, so the server can track which keys are currently down and interpret those as a combination.